contentprovider.go

```diff
@@ -454,6 +454,18 @@ func (nls newlines) getLines(data []byte, low, high int) []byte {
 	lowStart, _ := nls.lineBounds(low)
 	_, highEnd := nls.lineBounds(high - 1)
 
+	// Drop any trailing newline. Editors do not treat a trailing newline as
+	// the start of a new line, so we should not either. lineBounds clamps to
+	// len(data) when an out-of-bounds line is requested.
+	//
+	// As an example, if we request lines 1-5 from a file with contents
+	// `one\ntwo\nthree\n`, we should return `one\ntwo\nthree` because those are
+	// the three "lines" in the file, separated by newlines.
+	if highEnd == uint32(len(data)) && len(data) > 0 && data[len(data)-1] == '\n' {
```

Comment:
minor: ignore given that this may be faster since it avoids the function call:

```suggestion
	if highEnd == uint32(len(data)) && bytes.HasSuffix(data, []byte{'\n'}) {
```

Reply:
oh, way easier. This shouldn't be in a hot path anyways since since this condition will usually get short-circuited.